File: src/ts/estimate.ts
/**
 * Heuristic token estimation.
 *
 * Rule: ~3 chars per token for mixed PT/EN/code — a well-calibrated
 * average that holds within ±15% for typical developer content.
 *
 * Not a replacement for a real tokenizer. When exact counts matter,
 * use the provider's tokenizer (tiktoken, claude-tokenizer, etc.).
 */
export function estimateTokens(text: string | null | undefined): number {
  if (!text) return 0;
  return Math.ceil(text.length / 3);
}

export function byteLength(value: unknown): number {
  const s = typeof value === 'string' ? value : JSON.stringify(value);
  return Buffer.byteLength(s, 'utf8');
}

export function estimateObjectTokens(obj: unknown): number {
  return estimateTokens(JSON.stringify(obj));
}

File: src/py/omni_token_economy/estimate.py
"""Heuristic token and byte estimation. ~3 chars per token for mixed PT/EN/code."""
from __future__ import annotations

import json
import math
from typing import Any


def estimate_tokens(text: str | None) -> int:
    """Estimate tokens: ceil(len / 3). Not a real tokenizer — good enough for budgeting."""
    if not text:
        return 0
    return math.ceil(len(text) / 3)


def byte_length(value: Any) -> int:
    """UTF-8 byte length of a value (stringified if not a string)."""
    s = value if isinstance(value, str) else json.dumps(value, ensure_ascii=False)
    return len(s.encode("utf-8"))


def estimate_object_tokens(obj: Any) -> int:
    """Estimate tokens for an arbitrary serializable object."""
    return estimate_tokens(json.dumps(obj, ensure_ascii=False))
